server/src/main/java/io/pravega/schemaregistry/service/SchemaRegistryService.java

```diff
@@ -0,0 +1,924 @@
+import io.pravega.schemaregistry.contract.data.GroupHistoryRecord;
+import io.pravega.schemaregistry.contract.data.GroupProperties;
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.contract.data.SchemaWithVersion;
+import io.pravega.schemaregistry.contract.data.SerializationFormat;
+import io.pravega.schemaregistry.contract.data.VersionInfo;
+import io.pravega.schemaregistry.exceptions.IncompatibleSchemaException;
+import io.pravega.schemaregistry.exceptions.PreconditionFailedException;
+import io.pravega.schemaregistry.exceptions.SerializationFormatMismatchException;
+import io.pravega.schemaregistry.rules.CompatibilityChecker;
+import io.pravega.schemaregistry.rules.CompatibilityCheckerFactory;
+import io.pravega.schemaregistry.storage.ContinuationToken;
+import io.pravega.schemaregistry.storage.SchemaStore;
+import io.pravega.schemaregistry.storage.StoreExceptions;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.Schema;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static io.pravega.schemaregistry.contract.data.BackwardAndForward.*;
+
+/**
+ * Schema registry service backend.
+ */
+@Slf4j
+public class SchemaRegistryService {
+    private static final Retry.RetryAndThrowConditionally RETRY = Retry.withExpBackoff(1, 2, Integer.MAX_VALUE, 100)
+                                                                       .retryWhen(x -> Exceptions.unwrap(x) instanceof StoreExceptions.WriteConflictException);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final VersionInfo EMPTY_VERSION = new VersionInfo("", -1, -1);
+
+    static {
+        OBJECT_MAPPER.configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);
+        OBJECT_MAPPER.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
+    }
+    
+    private final SchemaStore store;
+
+    private final ScheduledExecutorService executor;
+
+    public SchemaRegistryService(SchemaStore store, ScheduledExecutorService executor) {
+        this.store = store;
+        this.executor = executor;
+    }
+
+    /**
+     * Lists groups with pagination.
+     *
+     * @param namespace         namespace for which the request is scoped to.
+     * @param continuationToken continuation token.
+     * @param limit             max number of groups to return.
+     * @return CompletableFuture which holds map of groups names and group properties upon completion.
+     */
+    public CompletableFuture<MapWithToken<String, GroupProperties>> listGroups(String namespace, ContinuationToken continuationToken, int limit) {
+        log.info("List groups called");
```

Comment:
`trace` or `traceRequest`

Reply:
answered earlier. we want to log at info and not trace level. tracerequest adds trace level logs